### ISSUE TITLE
Extract path object into new class

### ISSUE THREAD

**Issue by TheDudeFromCI** [Karang/mineflayer-pathfinder#13]
Currently, the path object is represented as an array of movements within the index.js, tightly bound with the moving process. I would like to split this into a new class object to make it more reusable. For example, generating multiple paths without actually starting the move action (for example, deciding which of several different destinations would be the easiest to get to), or allowing post-processing to be performed on the path before starting it. (An example of this could be to trim redundant waypoints, or shift them off the grid to make them feel more natural)

I can write this as a PR but would like permission to do this before writing anything.

**Comment by Karang** [comment]
What would this class contains apart from an array of waypoint ?

For more natural motion, you may want to look into the "free motion" experimental feature: https://github.com/Karang/mineflayer-pathfinder/blob/master/index.js#L174 for now it is only used when near the target, but the plan is to use it along with some way-point trimming.

I think the first thing to implement is the first item on the readme.md todo list:
* Make computations span multiple ticks
This will shape the design of the path api better. (We need to be able to append way-points to an already existing path)

Way-point trimming and post-processing should be done when walking the path node by node (and not before), because if the path needs to be recomputed, we would have wasted some computation power. (Trimming can be compute intensive if it is done in a way that respect the physics, see https://github.com/Karang/mineflayer-pathfinder/blob/master/index.js#L150)

**Comment by TheDudeFromCI** [comment]
Alright, I'll see what I can do. Should be a very easy implementation. I will still extract the path for the sole purpose of making the code much easier to write, maintain, and reuse.

**Comment by TheDudeFromCI** [comment]
Okay, I've started work on the new changes. I'm making waypoints completely modular, so you can add new waypoint tasks with only a few lines of code. Such as climbing ladders or opening doors. Paths are not tightly bound with the bot as they are now, meaning it is extremely easy to span path generation out over several ticks or even generate multiple tasks at the same time.

My progress will be posted [here](https://github.com/TheDudeFromCI/mineflayer-pathfinder/tree/async-path-generation).

**Comment by TheDudeFromCI** [comment]
The project has started to deviate too much from the original codebase. Closing issue.

**Comment by Karang** [comment]
Can you open a draft Pull Request, so your progress is more easy to follow ? I have a few comments already, on https://github.com/TheDudeFromCI/mineflayer-pathfinder/commit/3cb3dd5f65636718d66a186beb34d68fd7aafbc4:
* Why keep a bot reference in each waypoint ? If it is not necessary, the waypoint should only contain field that represent it.
* What about waypoint where you need to place and break blocks before moving ? The inheritance solution you propose seems to not cover that case. In order to keep the code more simple, I suggest to keep only one class for representing a waypoint (no inheritance).
* The tick monitoring is also responsible to check if a path should be updated, I think it is better to keep it outside the path class (also because you don't want to constantly add/remove the listener)

Otherwise it looks good. Having a waypoint class instead of object literals for representing the path nodes should also help a little with the performances (make sure you integrate that in the movement generator + astar)

**Comment by TheDudeFromCI** [comment]
After several hours behind the code editor, I feel my edits with the project do not align with your original plans for pathfinder. As such, I don't plan on merging the project back into the repo and will probably just keep my edits for private usage, only. That's why I closed this issue after my ping on Discord.

![Screenshot from 2020-06-24 06-38-04](https://user-images.githubusercontent.com/3118172/85567184-81521580-b5e5-11ea-9c27-cd8049687f63.png)

I feel that my usage of pathfinder and required functions differ greatly from the intended usage of the original project. I have a lot more code that I haven't pushed yet that further abstracts and moves the project into an OOP style setup. I don't want to mess up your project, and I believe that's what I'll be doing if I continue.

I just don't work well when contributing to other people's projects. I sincerely apologize.
